File: frontend/js/proposal-fonts.js
// The proposal's own typeface for the Proposal Review editor, loaded from behind the login.
/*
 * Hanz, 2026-09-26: "why are the fonts and font sizes still not fixed?" The page's CSS (.tw-page in
 * styles.css) and every run of the .docx name "Zetta Serif Book", but nothing ever LOADED it: there
 * was no @font-face anywhere, so on any machine without the font installed the editor drew Georgia
 * while the PDF (LibreOffice in the container, font mounted from the host) printed Zetta Serif.
 * Georgia is wider, so the editor also wrapped, shrank and clipped boxes the PDF does not.
 *
 * LICENSED, SO THE APP SERVES IT ONLY BEHIND THE LOGIN, and neither git nor the image carries it:
 * the server mounts it (backend/proposal_fonts.py, whose docstring also says what that leaves open).
 * The files come from /api/proposal-font/<name>, which the same auth middleware as every staff
 * /api route gates (backend/proposal_fonts.py). They are
 * fetched here with the staff bearer token and handed to the FontFace API as BYTES:
 *   * a CSS url() font load cannot carry the Authorization header, so it would 401;
 *   * a blob: URL would be refused by nginx's CSP, which says `font-src 'self'`;
 *   * a FontFace built from an ArrayBuffer makes no fetch of its own, so font-src never applies.
 * The one request this file makes is a same-origin fetch, which `connect-src 'self'` allows.
 *
 * FAMILY NAMES AND DESCRIPTORS ARE THE FILES' OWN, read out of their `name` and `OS/2` tables:
 *   Zetta Serif-Book.otf   family (nameID 1, Windows) "Zetta Serif Book", usWeightClass 345, upright
 *   Zetta Serif.otf        family (nameID 1, Windows) "Zetta Serif",      usWeightClass 400, upright
 * "Zetta Serif Book" is the name the .docx runs carry and what LibreOffice's fontconfig matches.
 * Neither file has a bold or an italic face: LibreOffice synthesises the bold lead-ins, and so does
 * the browser (font-synthesis defaults to weight and style) as long as the face is registered as
 * what it really is. Registering it as bold would switch the synthesis off and print lead-ins in
 * the regular weight. test_proposal_fonts.py reads both tables out of the binaries and checks this.
 *
 * WHEN IT IS IN, the page re-measures: proposal-review.js registers its refit with whenLoaded(),
 * which runs it ONCE, after both faces have been added (or after the one that could be).
 *
 * FAILURE IS SILENT TO THE USER. Offline, a 401 from a lapsed session, bytes the browser refuses:
 * the page keeps the CSS fallback exactly as it drew before this file existed, the refit never
 * runs, and the console gets ONE warning.
 */
(function (root) {
  "use strict";

  // Bump when a font file on the server changes. The responses are cached for a year
  // (Cache-Control: private, immutable), so this query string is the only thing that moves a
  // browser off the old bytes. test_proposal_fonts.py pins the files' hashes against it.
  const VERSION = "1";

  const FACES = [
    { name: "zetta-serif-book", family: "Zetta Serif Book",
      descriptors: { style: "normal", weight: "345" } },
    { name: "zetta-serif", family: "Zetta Serif",
      descriptors: { style: "normal", weight: "400" } },
  ];

  let phase = "idle";          // idle -> loading -> loaded | failed
  let loading = null;          // the one load, shared by every caller
  const waiting = [];

  function warn(msg) {
    try { root.console.warn("[proposal font] " + msg + " The editor is using its fallback serif."); }
    catch (_) { /* no console: nothing to say it to */ }
  }

  function reason(err) {
    return err && err.message ? err.message : String(err);
  }

  function apiBase() {
    try {
      if (root.TW && typeof root.TW.resolveApiBase === "function") return root.TW.resolveApiBase() || "";
    } catch (_) { /* fall through to same-origin */ }
    return "";
  }

  function token() {
    try {
      return root.TWAuth && typeof root.TWAuth.token === "function" ? root.TWAuth.token() : null;
    } catch (_) { return null; }
  }

  // Called as methods of `root` (root.fetch, root.document.fonts.add), never detached: a detached
  // window.fetch throws "Illegal invocation" in Chromium.
  async function loadFace(face, tok) {
    const res = await root.fetch(apiBase() + "/api/proposal-font/" + face.name + "?v=" + VERSION,
                                 { headers: { Authorization: "Bearer " + tok },
                                   credentials: "same-origin" });
    if (!res || !res.ok) throw new Error("HTTP " + (res ? res.status : "no response"));
    const ff = new root.FontFace(face.family, await res.arrayBuffer(),
                                 Object.assign({}, face.descriptors));
    await ff.load();
    root.document.fonts.add(ff);
  }

  async function loadAll() {
    // tokenReady, not ready: this only wants to START FETCHING, and the token exists a whole /api/me
    // round trip before `ready` settles (see auth.js). Neither settles without a signed-in
    // @wetreadwell.com session, and then there is nothing to fetch with anyway.
    const auth = root.TWAuth;
    const gate = auth && (auth.tokenReady || auth.ready);
    if (gate && typeof gate.then === "function") await gate;
    if (typeof root.FontFace !== "function" || !root.document || !root.document.fonts
        || typeof root.fetch !== "function") {
      throw new Error("This browser cannot register a font from bytes.");
    }
    const tok = token();
    if (!tok) throw new Error("No sign-in token.");
    const failed = (await Promise.all(FACES.map(face =>
      loadFace(face, tok).then(() => null, err => face.family + " (" + reason(err) + ")"))))
      .filter(Boolean);
    if (failed.length) warn("Not loaded: " + failed.join(", ") + ".");
    return failed.length < FACES.length;
  }

  function run(fn) {
    try { fn(); } catch (err) {
      try { root.console.error("[proposal font] refit after the font loaded failed:", err); } catch (_) {}
    }
  }

  /** Start the load (once). Resolves true when at least one face is in document.fonts. */
  function start() {
    if (!loading) {
      phase = "loading";
      loading = loadAll().catch(err => { warn(reason(err)); return false; }).then(ok => {
        phase = ok ? "loaded" : "failed";
        const due = waiting.splice(0);
        if (ok) due.forEach(run);
        return ok;
      });
    }
    return loading;
  }

  /** Run `fn` once, after the faces are registered. Never runs it if the load failed. Starts the
   *  load if nothing has yet. */
  function whenLoaded(fn) {
    if (typeof fn === "function") {
      if (phase === "loaded") Promise.resolve().then(() => run(fn));
      else if (phase !== "failed") waiting.push(fn);
    }
    return start();
  }

  root.TWProposalFonts = { VERSION: VERSION, FACES: FACES, start: start, whenLoaded: whenLoaded };
})(typeof window !== "undefined" ? window : this);

File: backend/proposal_fonts.py
"""Treadwell's licensed proposal typeface: where the server finds it, and the route that hands it
to signed-in staff.

Hanz, 2026-09-26: "why are the fonts and font sizes still not fixed?" The proposal templates are
typeset in Zetta Serif, and the PDF prints in it (LibreOffice in the container, which finds fonts
through fontconfig). The Proposal Review editor named the same family in its CSS but never LOADED
it, so on every machine without the font installed it drew Georgia instead. Georgia is wider, so
the editor also wrapped, shrank and clipped boxes that the PDF does not.

WHERE THE FILES COME FROM. Zetta Serif is a LICENSED font. Hanz, 2026-09-26: "Move font off
GitHub." Since then neither the repository (public) nor the image (pushed to GHCR) carries it: the
SERVER supplies it at runtime.
  * On the VPS the two files sit in /opt/treadwell-fonts (HOST_DIR), and both compose files
    bind-mount that directory read-only onto MOUNT_DIR, which the Dockerfile leaves empty for it.
    MOUNT_DIR is under /usr/share/fonts, so LibreOffice's fontconfig finds the same files this
    route serves. fontconfig rescans a directory whose mtime no longer matches its cache, and the
    image's cache records MOUNT_DIR empty at build time, so nothing has to run fc-cache at start.
  * On a dev box they sit in backend/fonts/, which .gitignore keeps out of git and .dockerignore
    keeps out of any image built from the working tree (deploy/ship.sh builds from it). In the
    image that directory is /app/fonts and holds only its README.
  * The lasting copy is neither of those: it is the team Dropbox folder that backend/fonts/README.md
    names, with the sha256 of each file. Git deletes the files from every checkout that moves past
    the commit that untracked them, so each deploy path copies them from its own VPS checkout into
    HOST_DIR before its first such pull, and a new box or dev checkout copies them from Dropbox.

A MISSING FILE IS NOT AN ERROR ANYWHERE, and that is the danger. Docker creates a missing host
directory as an EMPTY one, so a box without the files boots, passes its healthcheck and prints
every PDF in a substitute font. So `report_once` says so, loudly, once per process, from a startup
hook in main.py. /healthz does not look: a liveness probe has to stay cheap and must never flap on
a font. The deploy workflow refuses to deploy while either file is missing from HOST_DIR, and the
route below answers 404 for a file that is not there, which the editor takes quietly (it keeps its
fallback serif). A file placed after start is picked up without a restart: only a hit is cached.

THE ROUTE. Hanz approved serving the font to staff from the app on the condition that it stays
behind the login. GET /api/proposal-font/{name} (main.py) is the only way the app hands the files
out, and it keeps to that condition:
  * it is served from an /api route, which puts it behind the Supabase auth middleware every other
    staff route sits behind (`_auth_gate` in main.py) -- an unauthenticated request is refused
    before routing, so no URL of THIS APP returns these bytes to a signed-out caller;
  * neither directory it reads is under the frontend/ directory the static mount serves;
  * the request names a font by a fixed PUBLIC NAME, and that name is looked up in `FONTS` below.
    Nothing a request carries is ever joined onto a path: a name that is not a key there resolves
    to nothing and the filesystem is never touched.

The browser side is frontend/js/proposal-fonts.js, which fetches both files with the staff bearer
token and registers them through the FontFace API.

WHAT THIS DOES NOT UNDO. The files were tracked in git until 2026-09-26, so the public
repository's HISTORY still holds them, and every image built before that date carries
them (at /usr/share/fonts/truetype/treadwell and /app/fonts) in GHCR. Making the GHCR package
private and rewriting the history are repository and registry decisions for Hanz; nothing in this
file can do either.
"""
import hashlib
import logging
import pathlib
import threading
from typing import Dict, List, Optional, Tuple

# Public name -> file name. The ONLY route from a request to a file.
FONTS: Dict[str, str] = {
    "zetta-serif-book": "Zetta Serif-Book.otf",
    "zetta-serif": "Zetta Serif.otf",
}

MEDIA_TYPE = "font/otf"

# `private` because the response is for one signed-in user: no shared cache (a proxy, a CDN) may
# keep a copy and hand it to somebody who is not signed in. A year and `immutable` because the file
# never changes under a given URL -- frontend/js/proposal-fonts.js puts a VERSION in the query string
# and bumps it when a file changes (test_proposal_fonts.py pins the hashes to catch that).
CACHE_CONTROL = "private, max-age=31536000, immutable"

_HERE = pathlib.Path(__file__).resolve().parent

# The runtime mount point inside the container. Both compose files mount HOST_DIR here read-only.
MOUNT_DIR = pathlib.Path("/usr/share/fonts/truetype/treadwell")
# Where the files go on the VPS. The app never reads it (it is outside the container); it is here
# so the startup warning can say where to put them.
HOST_DIR = "/opt/treadwell-fonts"

# Where the files are read from, in order: the mount, then backend/fonts/ for local dev.
FONT_DIRS: Tuple[pathlib.Path, ...] = (
    MOUNT_DIR,
    _HERE / "fonts",
)

_cache: Dict[str, Tuple[bytes, str]] = {}
_lock = threading.Lock()
_reported = False


def load(name: object) -> Optional[Tuple[bytes, str]]:
    """(bytes, etag) for a whitelisted public font name, or None.

    None for a name that is not in `FONTS` -- decided before any path is built -- and for a
    whitelisted font whose file is missing from every directory in `FONT_DIRS`. The bytes are read
    once per process and kept: two files, about 190 KB between them."""
    if not isinstance(name, str):
        return None
    filename = FONTS.get(name)
    if filename is None:
        return None
    with _lock:
        hit = _cache.get(name)
        if hit is not None:
            return hit
        for folder in FONT_DIRS:
            try:
                data = (folder / filename).read_bytes()
            except OSError:
                continue
            etag = '"' + hashlib.sha256(data).hexdigest()[:32] + '"'
            _cache[name] = (data, etag)
            return _cache[name]
    return None


def cache_clear() -> None:
    """Forget what has been read (tests that move the files use this)."""
    with _lock:
        _cache.clear()


def missing() -> List[str]:
    """The file names in `FONTS` found in none of `FONT_DIRS`, in `FONTS` order."""
    return [filename for filename in FONTS.values()
            if not any((folder / filename).is_file() for folder in FONT_DIRS)]


def report_once(logger: logging.Logger) -> bool:
    """Log whether the proposal font is installed, once per process. True if this call logged.

    One INFO line when both files are there. Otherwise one WARNING that names what is missing,
    where the app looked, what that costs and where the files go -- written to be found by
    `docker compose logs | grep -i font` on a box that looks perfectly healthy."""
    global _reported
    with _lock:
        if _reported:
            return False
        _reported = True
    gone = missing()
    looked = " and ".join(folder.as_posix() for folder in FONT_DIRS)
    if not gone:
        logger.info("Proposal font: Zetta Serif is installed (%s).",
                    ", ".join(sorted(FONTS.values())))
        return True
    which = ("neither Zetta Serif file (%s) is" % ", ".join(gone)
             if len(gone) == len(FONTS) else "%s is" % ", ".join(gone))
    logger.warning(
        "PROPOSAL FONT MISSING: %s in %s. Proposal PDFs will print in a substitute font and the "
        "editor falls back to Georgia. Put %s in %s on the host (compose mounts it read-only at "
        "%s), then restart the container.",
        which, looked, " and ".join('"%s"' % f for f in FONTS.values()), HOST_DIR,
        MOUNT_DIR.as_posix())
    return True
